File: src/webassembly/assembly/placing-four-sides.ts
import { AxisAlignedBoundingBox, QuadTree, Vector2 } from './quadtree';
import {
  isZero,
  NULL_RANGE,
  PositionInfo,
  Range,
  TRangeSet,
  WordCloudTopic,
} from './word-cloud-placing';

export function calcMinMax(dir: Vector2, points: StaticArray<Vector2>): Range {
  let max = -f32.MAX_VALUE;
  let min = f32.MAX_VALUE;
  for (let i = 0; i < points.length; i++) {
    const v = dir.dot(points[i]);
    if (max < v) max = v;
    if (min > v) min = v;
  }
  return new Range(min, max);
}

class SimpleCollisionBox {
  private readonly points: StaticArray<Vector2> = new StaticArray<Vector2>(4);
  private readonly completeMoveRange: Range;
  private readonly collRange: Range;
  private readonly moveRange: Range;
  private readonly xRange: Range;
  private readonly yRange: Range;
  private resultTRange: TRangeSet;

  /**
   * @param midPoint Currently testing mid point
   * @param moveDir direction of movement for t (is perpendicular to one side)
   * @param addCollDir direction of second normal (perpendicular to move)
   * @param moveSize size of normal box along move dir
   * @param addCollSize size of normal box along second normal
   * @param moveTSize size of t range (at least moveSize)
   */
  constructor(
    public readonly midPoint: Vector2,
    public readonly moveDir: Vector2,
    public readonly addCollDir: Vector2,
    private readonly moveSize: f32,
    private readonly addCollSize: f32,
    private readonly rotation: f32,
    moveTSize: f32
  ) {
    this.points[0] = moveDir
      .clone()
      .scale(moveSize)
      .add(addCollDir.clone().scale(addCollSize))
      .add(midPoint);
    this.points[1] = moveDir
      .clone()
      .scale(moveSize)
      .add(addCollDir.clone().scale(-addCollSize))
      .add(midPoint);
    this.points[2] = moveDir
      .clone()
      .scale(-moveSize)
      .add(addCollDir.clone().scale(addCollSize))
      .add(midPoint);
    this.points[3] = moveDir
      .clone()
      .scale(-moveSize)
      .add(addCollDir.clone().scale(-addCollSize))
      .add(midPoint);
    this.moveRange = calcMinMax(this.moveDir, this.points);
    this.completeMoveRange = new Range(
      this.moveRange.start - moveTSize,
      this.moveRange.end + moveTSize
    );
    this.collRange = calcMinMax(this.addCollDir, this.points);
    this.resultTRange = new TRangeSet(moveTSize);
    // calculate for sat
    const edgePoints = new StaticArray<Vector2>(4);
    {
      const fullyLength = moveSize + moveTSize;
      edgePoints[0] = this.moveDir
        .clone()
        .scale(fullyLength)
        .add(this.addCollDir.clone().scale(addCollSize))
        .add(this.midPoint);
      edgePoints[1] = this.moveDir
        .clone()
        .scale(fullyLength)
        .add(this.addCollDir.clone().scale(-addCollSize))
        .add(this.midPoint);
      edgePoints[2] = this.moveDir
        .clone()
        .scale(-fullyLength)
        .add(this.addCollDir.clone().scale(addCollSize))
        .add(this.midPoint);
      edgePoints[3] = this.moveDir
        .clone()
        .scale(-fullyLength)
        .add(this.addCollDir.clone().scale(-addCollSize))
        .add(this.midPoint);
    }
    this.xRange = calcMinMax(new Vector2(1, 0), edgePoints);
    this.yRange = calcMinMax(new Vector2(0, 1), edgePoints);
  }

  /**
   * @param aspectRatio width / height from screen
   */
  constructPosition(aspectRatio: f32): PositionInfo | null {
    if (this.resultTRange.isEmpty()) return null;
    /*
    get minimal distance with using perpendicular side
    from equality (distance to center):
     (0 0) + k * (c1 c2) = (m1 m2) + t * (move1 move2)
    */
    const c1 = this.addCollDir.getDirectionX() / aspectRatio;
    const c2 = this.addCollDir.getDirectionY();
    const m1 = this.midPoint.getDirectionX() / aspectRatio;
    const m2 = this.midPoint.getDirectionY();
    const move1 = this.moveDir.getDirectionX() / aspectRatio;
    const move2 = this.moveDir.getDirectionY();
    const optimalT = (c2 * m1 - c1 * m2) / (c1 * move2 - c2 * move1);
    // find possible t
    const t = this.resultTRange.getBestTValue(optimalT);
    // construct position
    const midPoint = this.moveDir.clone().scale(t).add(this.midPoint);
    return new PositionInfo(
      midPoint,
      midPoint.lengthSquared(aspectRatio),
      this.moveDir,
      this.addCollDir,
      this.moveSize,
      this.addCollSize
    );
  }

  /**
   * @returns
   * true, if the collision checks can continue
   *
   * false, if it has no space, collision checks can be aborted
   */
  collideTRange(topic: WordCloudTopic): bool {
    const pos = topic.position!;
    // self move dir
    const selfDir1 = this.resultTRange.mergeTRange(
      calcMinMax(this.moveDir, pos.points),
      this.moveRange,
      1
    );
    if (selfDir1 === NULL_RANGE) {
      return true;
    }
    // self additional dir
    const selfDir2 = this.resultTRange.mergeTRange(
      calcMinMax(this.addCollDir, pos.points),
      this.collRange,
      0
    );
    if (selfDir2 === NULL_RANGE) {
      return true;
    }
    const rotation = this.rotation - topic.rotation;
    if (isZero(rotation % 90)) {
      if (selfDir1 !== null) {
        this.resultTRange.splitByRange(selfDir1.start, selfDir1.end);
      }
      if (selfDir2 !== null) {
        this.resultTRange.splitByRange(selfDir2.start, selfDir2.end);
      }
      return !this.resultTRange.isEmpty();
    }
    // other up
    const otherDir1 = this.resultTRange.mergeTRange(
      pos.normal1Range,
      calcMinMax(pos.normal1, this.points),
      pos.normal1.dot(this.moveDir)
    );
    if (otherDir1 === NULL_RANGE) {
      return true;
    }
    // other right
    const otherDir2 = this.resultTRange.mergeTRange(
      pos.normal2Range,
      calcMinMax(pos.normal2, this.points),
      pos.normal2.dot(this.moveDir)
    );
    if (otherDir2 === NULL_RANGE) {
      return true;
    }
    // update possible t ranges
    if (selfDir1 !== null) {
      this.resultTRange.splitByRange(selfDir1.start, selfDir1.end);
    }
    if (selfDir2 !== null) {
      this.resultTRange.splitByRange(selfDir2.start, selfDir2.end);
    }
    if (otherDir1 !== null) {
      this.resultTRange.splitByRange(otherDir1.start, otherDir1.end);
    }
    if (otherDir2 !== null) {
      this.resultTRange.splitByRange(otherDir2.start, otherDir2.end);
    }
    return !this.resultTRange.isEmpty();
  }

  collideSAT(box: AxisAlignedBoundingBox<f32>): bool {
    // box up
    if (this.yRange.start >= box.y + box.height || this.yRange.end <= box.y) {
      return false;
    }
    // box right
    if (this.xRange.start >= box.x + box.width || this.xRange.end <= box.x) {
      return false;
    }
    // self up
    if (
      !this.completeMoveRange.collides(calcMinMax(this.moveDir, box.points))
    ) {
      return false;
    }
    // self right
    if (!this.collRange.collides(calcMinMax(this.addCollDir, box.points))) {
      return false;
    }
    return true;
  }
}

export const tryPlaceOnFourSides = (
  topic: WordCloudTopic,
  newTopic: WordCloudTopic,
  quadTree: QuadTree<f32, WordCloudTopic>,
  aspectRatio: f32
): PositionInfo | null => {
  const rotation: f32 = newTopic.rotation - topic.rotation;
  let rotIndex: i32 = i32(
    ((rotation < 0 ? rotation + 405 : rotation + 45) / 90) % 4
  );
  let newSideIndex = (rotIndex + 2) % 4;
  let minimal: PositionInfo | null = null;
  for (let i = 0; i < 4; ++i, newSideIndex = (newSideIndex + 1) % 4) {
    const newSide = newTopic.buildInfo.sides[newSideIndex];
    const side = topic.buildInfo.sides[i];
    const currentMid = topic.position!.mid;
    const newMid = side.sideNormal
      .clone()
      .scale(side.distToMid + newSide.distToMid)
      .add(currentMid);
    const collisionBox = new SimpleCollisionBox(
      newMid,
      side.perpendicularNormal.clone(),
      side.sideNormal.clone(),
      newSide.otherDirMidDist,
      newSide.distToMid,
      newTopic.rotation,
      side.otherDirMidDist + newSide.otherDirMidDist
    );
    quadTree.queryObjects(
      collisionBox,
      (alignedBox, box) => box.collideSAT(alignedBox),
      (object, topics) => {
        for (let j = 0; j < topics.length; j++) {
          if (!object.collideTRange(topics[j])) {
            return false;
          }
        }
        return true;
      }
    );
    const t = collisionBox.constructPosition(aspectRatio);
    if (t == null) {
      continue;
    }
    if (minimal == null) {
      minimal = t;
    } else if (minimal.distanceSquared > t.distanceSquared) {
      minimal = t;
    }
  }
  return minimal;
};

File: webassembly/frag-jetzt-wasm/src/wordCloudPlacing/placingFourSides.rs
use std::{cell::RefCell, rc::Rc};

use super::{
    quadtree::{
        calcMinMax, AxisAlignedBoundingBox, Number, QuadTree, QuadTreeQuery, Range, Vector2,
    },
    MergeRangeResult, PositionInfo, TRangeSet, WordCloudElement, WordCloudTopic,
};

struct SimpleCollisionBox<T: Number> {
    midPoint: Vector2<T>,
    moveDir: Vector2<T>,
    addCollDir: Vector2<T>,
    moveSize: T,
    addCollSize: T,
    rotation: T,
    points: Vec<Vector2<T>>,
    completeMoveRange: Range<T>,
    collRange: Range<T>,
    moveRange: Range<T>,
    xRange: Range<T>,
    yRange: Range<T>,
    resultTRange: TRangeSet<T>,
}

impl<T: Number> SimpleCollisionBox<T> {
    pub fn new(
        midPoint: &Vector2<T>,
        moveDir: &Vector2<T>,
        addCollDir: &Vector2<T>,
        moveSize: T,
        addCollSize: T,
        rotation: T,
        moveTSize: T,
    ) -> Self {
        let points = vec![
            moveDir
                .clone()
                .scale(moveSize)
                .add(&addCollDir.clone().scale(addCollSize))
                .add(&midPoint)
                .clone(),
            moveDir
                .clone()
                .scale(moveSize)
                .add(&addCollDir.clone().scale(-addCollSize))
                .add(&midPoint)
                .clone(),
            moveDir
                .clone()
                .scale(-moveSize)
                .add(&addCollDir.clone().scale(addCollSize))
                .add(&midPoint)
                .clone(),
            moveDir
                .clone()
                .scale(-moveSize)
                .add(&addCollDir.clone().scale(-addCollSize))
                .add(&midPoint)
                .clone(),
        ];
        let moveRange = calcMinMax(&moveDir, &points);
        let xRange;
        let yRange;
        {
            let fullyLength = moveSize + moveTSize;
            let edgePoints = [
                moveDir
                    .clone()
                    .scale(fullyLength)
                    .add(&addCollDir.clone().scale(addCollSize))
                    .add(&midPoint)
                    .clone(),
                moveDir
                    .clone()
                    .scale(fullyLength)
                    .add(&addCollDir.clone().scale(-addCollSize))
                    .add(&midPoint)
                    .clone(),
                moveDir
                    .clone()
                    .scale(-fullyLength)
                    .add(&addCollDir.clone().scale(addCollSize))
                    .add(&midPoint)
                    .clone(),
                moveDir
                    .clone()
                    .scale(-fullyLength)
                    .add(&addCollDir.clone().scale(-addCollSize))
                    .add(&midPoint)
                    .clone(),
            ];
            xRange = calcMinMax(&Vector2::dir_x(), &edgePoints);
            yRange = calcMinMax(&Vector2::dir_y(), &edgePoints);
        }
        SimpleCollisionBox {
            midPoint: midPoint.clone(),
            moveDir: moveDir.clone(),
            collRange: calcMinMax(&addCollDir, &points),
            addCollDir: addCollDir.clone(),
            moveSize,
            addCollSize,
            rotation,
            points,
            completeMoveRange: Range::new(
                moveRange.get_start() - moveTSize,
                moveRange.get_end() + moveTSize,
            ),
            moveRange,
            xRange,
            yRange,
            resultTRange: TRangeSet::new(moveTSize),
        }
    }

    /**
     * @param aspectRatio width / height from screen
     */
    pub fn constructPosition(&self, aspectRatio: T) -> Option<PositionInfo<T>> {
        if self.resultTRange.isEmpty() {
            return None;
        }
        /*
        get minimal distance with using perpendicular side
        from equality (distance to center):
         (0 0) + k * (c1 c2) = (m1 m2) + t * (move1 move2)
        */
        let c1 = self.addCollDir.get_x() / aspectRatio;
        let c2 = self.addCollDir.get_y();
        let m1 = self.midPoint.get_x() / aspectRatio;
        let m2 = self.midPoint.get_y();
        let move1 = self.moveDir.get_x() / aspectRatio;
        let move2 = self.moveDir.get_y();
        let optimalT = (c2 * m1 - c1 * m2) / (c1 * move2 - c2 * move1);
        // find possible t
        let t = self.resultTRange.getBestTValue(optimalT);
        // construct position
        let midPoint = self.moveDir.clone().scale(t).add(&self.midPoint).clone();
        Some(PositionInfo::new(
            midPoint.clone(),
            midPoint.length_squared(aspectRatio),
            self.moveDir.clone(),
            self.addCollDir.clone(),
            self.moveSize,
            self.addCollSize,
        ))
    }

    /**
     * @returns
     * true, if the collision checks can continue
     *
     * false, if it has no space, collision checks can be aborted
     */
    fn collideTRange(&mut self, object: &Rc<RefCell<dyn WordCloudElement<T>>>) -> bool {
        let borrow = object.borrow();
        let (points, topicRotation, normal1, normal1Range, normal2, normal2Range) =
            borrow.get_collison_info();

        // self move dir
        let selfDir1 = self.resultTRange.mergeTRange(
            &calcMinMax(&self.moveDir, points),
            &self.moveRange,
            T::One(),
        );
        if selfDir1.never_collides() {
            return true;
        }
        // self additional dir
        let selfDir2 = self.resultTRange.mergeTRange(
            &calcMinMax(&self.addCollDir, points),
            &self.collRange,
            T::Zero(),
        );
        if selfDir2.never_collides() {
            return true;
        }
        let rotation = self.rotation - topicRotation;
        if T::isZero(rotation % T::Ninety()) {
            if let MergeRangeResult::CollisionRange(range) = selfDir1 {
                self.resultTRange
                    .splitByRange(range.get_start(), range.get_end());
            }
            if let MergeRangeResult::CollisionRange(range) = selfDir2 {
                self.resultTRange
                    .splitByRange(range.get_start(), range.get_end());
            }
            return !self.resultTRange.isEmpty();
        }
        // other up
        let otherDir1 = self.resultTRange.mergeTRange(
            normal1Range,
            &calcMinMax(&normal1, &self.points),
            normal1.dot(&self.moveDir),
        );
        if otherDir1.never_collides() {
            return true;
        }
        // other right
        let otherDir2 = self.resultTRange.mergeTRange(
            normal2Range,
            &calcMinMax(&normal2, &self.points),
            normal2.dot(&self.moveDir),
        );
        if otherDir2.never_collides() {
            return true;
        }
        // update possible t ranges
        if let MergeRangeResult::CollisionRange(range) = selfDir1 {
            self.resultTRange
                .splitByRange(range.get_start(), range.get_end());
        }
        if let MergeRangeResult::CollisionRange(range) = selfDir2 {
            self.resultTRange
                .splitByRange(range.get_start(), range.get_end());
        }
        if let MergeRangeResult::CollisionRange(range) = otherDir1 {
            self.resultTRange
                .splitByRange(range.get_start(), range.get_end());
        }
        if let MergeRangeResult::CollisionRange(range) = otherDir2 {
            self.resultTRange
                .splitByRange(range.get_start(), range.get_end());
        }
        return !self.resultTRange.isEmpty();
    }
}

impl<T: Number> QuadTreeQuery<T> for SimpleCollisionBox<T> {
    fn collidesWithTree(&self, aabb: &AxisAlignedBoundingBox<T>) -> bool {
        self.yRange.get_start() < aabb.y + aabb.height
            && self.yRange.get_end() > aabb.y
            && self.xRange.get_start() < aabb.x + aabb.width
            && self.xRange.get_end() > aabb.x
            && self
                .completeMoveRange
                .collides(&calcMinMax(&self.moveDir, &aabb.points))
            && self
                .collRange
                .collides(&calcMinMax(&self.addCollDir, &aabb.points))
    }

    fn shouldContinueColliding(
        &mut self,
        objects: &[Rc<RefCell<dyn WordCloudElement<T>>>],
    ) -> bool {
        for object in objects {
            if !self.collideTRange(object) {
                return false;
            }
        }
        true
    }
}

pub fn tryPlaceOnFourSides<T: Number>(
    topic: &WordCloudTopic<T>,
    newTopic: &WordCloudTopic<T>,
    quadTree: &QuadTree<T>,
    aspectRatio: T,
) -> Option<PositionInfo<T>> {
    let rotation = newTopic.rotation - topic.rotation;
    let rotIndex = rotation.to_rotation_index();
    let mut newSideIndex = (rotIndex + 2) % 4;
    let mut minimal: Option<PositionInfo<T>> = None;
    for i in 0..4 {
        let newSide = &newTopic.buildInfo.sides[newSideIndex];
        newSideIndex = (newSideIndex + 1) % 4;
        let side = &topic.buildInfo.sides[i];
        let currentMid = &topic.place.as_ref().unwrap().mid;
        let newMid = side
            .sideNormal
            .clone()
            .scale(side.distToMid + newSide.distToMid)
            .add(&currentMid)
            .clone();
        let mut collisionBox = SimpleCollisionBox::new(
            &newMid,
            &side.perpendicularNormal,
            &side.sideNormal,
            newSide.otherDirMidDist,
            newSide.distToMid,
            newTopic.rotation,
            side.otherDirMidDist + newSide.otherDirMidDist,
        );
        quadTree.queryObjects(&mut collisionBox);
        let t = collisionBox.constructPosition(aspectRatio);
        if t.is_none() {
            continue;
        }
        if minimal.is_none() {
            minimal = t;
        } else if minimal.as_ref().unwrap().distanceSquared > t.as_ref().unwrap().distanceSquared {
            minimal = t;
        }
    }
    minimal
}
